exchange/bitswap/decision/engine.go

```diff
@@ -298,6 +298,27 @@ func (e *Engine) findOrCreate(p peer.ID) *ledger {
 	return l
 }
 
+type LedgerSnapshot struct {
+	DebtRatio     float64
+	BytesSent     uint64
+	BytesReceived uint64
+}
+
+func (e *Engine) LedgerSnapshot(p peer.ID) *LedgerSnapshot {
+	e.lock.Lock()
+	defer e.lock.Unlock()
```

Comment:
This will deadlock. `findOrCreate` takes the same lock that youre taking here. The correct way to do it will be to take the lock on the ledger you get back.